Task2.2/Program.cs

```diff
@@ -0,0 +1,159 @@
+*/
+
+class Program
+{
+    public class DiagonalMatrix
+    {
+        private readonly int[] diagonalElements;
+
+        // readonly property to get the size of the matrix
+        public int Size { get; }
+
+        // constructor
+        public DiagonalMatrix(params int[] elements)
+        {
+            if (elements == null)
+            {
+                Size = 0;
+                diagonalElements = new int[0];
+            }
+            else
+            {
+                Size = elements.Length;
+                diagonalElements = new int[Size];
+                for (int i = 0; i < Size; i++)
+                {
+                    diagonalElements[i] = elements[i];
+                }
+            }
+        }
+
+        // indexer
+        public int this[int i, int j]
+        {
+            get
+            {
+                if (i < 0 || i >= Size || j < 0 || j >= Size || j != i)
+                {
+                    return 0; //I don't know which option is preferd given different versions in task description and additional note below
+                    // throw new ArgumentOutOfRangeException("Indices are out of bounds.");
+                }
+                return diagonalElements[i];
+            }
+            set
+            {
+                if (i < 0 || i >= Size || j < 0 || j >= Size || i != j)
+                {
+                    throw new ArgumentOutOfRangeException("Indices are out of bounds.");
+                }
+                diagonalElements[i] = value;
+            }
+        }
+
+        // method to sum elements on diagonal (Track)
+        public int Track()
+        {
+            int sum = 0;
+            for (int i = 0; i < Size; i++)
+            {
+                sum += diagonalElements[i];
+            }
+            return sum;
+        }
+        public override bool Equals(object obj)
+        {
+            if (obj is DiagonalMatrix otherMatrix)
+            {
+                if (Size != otherMatrix.Size)
+                {
+                    return false;
+                }
+
+                for (int i = 0; i < Size; i++)
+                {
+                    if (diagonalElements[i] != otherMatrix.diagonalElements[i])
+                    {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            return false;
+        }
+        public override string ToString()
+        {
+            if (Size == 0) return "[]";
+
+            string result = "[ ";
+            for (int i = 0; i < Size; i++)
+            {
+                result += diagonalElements[i] + " ";
+            }
+            result += "]";
+            return result;
+        }
+
+        // Adds matrices treating missing elements as 0's
```

Comment:
It has to be extension method . I will explain on lectures later, but now can read 
https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/extension-methods